crates/oxc_linter/src/rules/typescript/prefer_for_of.rs

```diff
@@ -0,0 +1,296 @@
+    /// }
+    ///
+    /// // Good
+    /// for (const item of arr) {
+    ///   console.log(item);
+    /// }
+    /// ```
+    PreferForOf,
+    correctness
+);
+
+fn is_assignee(node: &AstNode) -> bool {
+    match node.kind() {
+        AstKind::SimpleAssignmentTarget(_)
+        | AstKind::UpdateExpression(_)
+        | AstKind::ArrayPattern(_) => true,
+        AstKind::UnaryExpression(unary_expr) => unary_expr.operator == UnaryOperator::Delete,
+        _ => false,
+    }
+}
+
+impl Rule for PreferForOf {
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        let AstKind::ForStatement(for_stmt) = node.kind() else { return };
+
+        let Some(ForStatementInit::VariableDeclaration(for_stmt_init)) = &for_stmt.init else {
+            return;
+        };
+
+        if for_stmt_init.declarations.len() != 1
+            || for_stmt_init.kind == VariableDeclarationKind::Const
+        {
+            return;
+        }
+
+        let decl = &for_stmt_init.declarations[0];
+        let (var_name, var_symbol_id) = match &decl.id.kind {
+            BindingPatternKind::BindingIdentifier(id) => (&id.name, id.symbol_id.get()),
+            _ => return,
+        };
+
+        if !matches!(&decl.init,
+            Some(Expression::NumericLiteral(literal)) if literal.value == 0f64
+        ) {
+            return;
+        }
+
+        let Some(Expression::BinaryExpression(test_expr)) = &for_stmt.test else { return };
+
+        if !matches!((&test_expr.left, test_expr.operator),
+            (Expression::Identifier(id), BinaryOperator::LessThan) if id.name == var_name
+        ) {
+            return;
+        }
+
+        let array_name = {
+            let Expression::MemberExpression(mem_expr) = &test_expr.right else { return };
+            if !matches!(mem_expr.static_property_name(), Some(prop_name) if prop_name == "length")
+            {
+                return;
+            }
+
+            match &mem_expr.object() {
+                Expression::Identifier(id) => id.name.as_str(),
+                Expression::MemberExpression(m_expr) => match m_expr.static_property_name() {
+                    Some(array_name) => array_name,
+                    None => return,
+                },
+                _ => return,
+            }
+        };
+
+        let Some(update_expr) = &for_stmt.update else { return };
+        if !update_expr.is_increment_of(var_name) {
+            return;
+        }
+
+        let nodes = ctx.nodes();
+        let test_and_update_span = test_expr.span.merge(&update_expr.span());
+        let body_span = for_stmt.body.span();
+
+        let Some(var_symbol_id) = var_symbol_id else { return };
+        if ctx.semantic().symbol_references(var_symbol_id).any(|reference| {
+            let ref_id = reference.node_id();
+
+            let symbol_span = nodes.get_node(ref_id).kind().span();
+            if test_and_update_span.contains(symbol_span) || !body_span.contains(symbol_span) {
```

Comment:
Why do we need to check `test_and_update_span`?

Reply:
I can't. Probably something from the earlier versions. I deleted it.